wireguard/wireguard.go

```diff
@@ -0,0 +1,1395 @@
+			w.logCxt.Debug("Node updated")
+			w.setNode(name, node)
+		} else {
+			// No further update, delete update so it's not processed again.
+			w.logCxt.Debug("No updates for the node - remove node update to remove additional processing")
+			delete(w.nodeUpdates, name)
+		}
+	}
+}
+
+// updateRouteTable updates the route table from the node updates.
+func (w *Wireguard) updateRouteTableFromNodeUpdates() {
+	for name, update := range w.nodeUpdates {
+		w.logCxt.Debugf("Processing node %s", name)
+		node := w.getNode(name)
+
+		// Delete routes that are no longer required in routing.
+		update.allowedCidrsDeleted.Iter(func(item interface{}) error {
+			w.logCxt.Debugf("Removing CIDR %s from routetable", item)
+			cidr := item.(ip.CIDR)
+			w.routetable.RouteRemove(w.config.InterfaceName, cidr)
+			return nil
+		})
+
+		// If the node routing to wireguard does not match with whether we should route then we need to do a full
+		// route update, otherwise do an incremental update.
+		var updateSet set.Set
+		shouldRouteToWireguard := w.shouldRouteToWireguard(node)
+		if node.routingToWireguard != shouldRouteToWireguard {
+			w.logCxt.Debug("Wireguard routing has changed - need to update full set of CIDRs")
+			updateSet = node.cidrs
+		} else {
+			w.logCxt.Debug("Wireguard routing has changed - need to update added CIDRs")
+			updateSet = update.allowedCidrsAdded
+		}
+
+		var targetType routetable.TargetType
+		var ifaceName, deleteIfaceName string
+		if !shouldRouteToWireguard {
+			// If we should not route to wireguard then we need to use a throw directive to skip wireguard routing and
+			// return to normal routing. We may also need to delete the existing route to wireguard.
+			w.logCxt.Debug("Not routing to wireguard - set route type to throw")
+			targetType = routetable.TargetTypeThrow
+			ifaceName = routetable.InterfaceNone
+			deleteIfaceName = w.config.InterfaceName
+		} else {
+			// If we should route to wireguard then route to the wireguard interface. We may also need to delete the
+			// existing throw route that was used to circumvent wireguard routing.
+			w.logCxt.Debug("Routing to wireguard interface")
+			ifaceName = w.config.InterfaceName
+			deleteIfaceName = routetable.InterfaceNone
+		}
+
+		updateSet.Iter(func(item interface{}) error {
+			cidr := item.(ip.CIDR)
+			w.logCxt.Debugf("Updating route for CIDR %s", cidr)
+			if node.routingToWireguard != shouldRouteToWireguard {
+				// The wireguard setting has changed. It is possible that some of the entries we are "removing" were
+				// never added - the routetable component handles that gracefully. We need to do these deletes because
+				// routetable component groups by interface and we are essentially moving routes between the wireguard
+				// interface and the "none" interface.
+				w.logCxt.Debugf("Wireguard routing has changed - delete previous route for %s", deleteIfaceName)
+				w.routetable.RouteRemove(deleteIfaceName, cidr)
```

Comment:
When you change the interface, could you hit that problem we discussed yesterday:

- route table iterates over interfaces
- route table happens to pick the interface where routes were added first
- route table tries to add routes, fails (because they already exist elsewhere), spews errors
- route table updates other interface, does deletions
- route update is retried and all routes get programmed

We only hit that in rare corner cases for workloads (e.g. IP reuse), is this more mainline for wireguard?

Reply:
no less mainline - generally i'd expect the wireguard interface to be set up once and never changed.